server/src/main/java/io/druid/server/http/SegmentListerResource.java

```diff
@@ -0,0 +1,218 @@
+/*
+ * Licensed to Metamarkets Group Inc. (Metamarkets) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.server.http;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.jaxrs.smile.SmileMediaTypes;
+import com.google.common.util.concurrent.FutureCallback;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.inject.Inject;
+import com.metamx.emitter.EmittingLogger;
+import com.sun.jersey.spi.container.ResourceFilters;
+import io.druid.guice.annotations.Json;
+import io.druid.guice.annotations.Smile;
+import io.druid.server.coordination.DataSegmentAnnouncer;
+import io.druid.server.coordination.SegmentChangeRequestHistory;
+import io.druid.server.coordination.SegmentChangeRequestsSnapshot;
+import io.druid.server.http.security.StateResourceFilter;
+import io.druid.server.security.AuthConfig;
+
+import javax.servlet.AsyncContext;
+import javax.servlet.AsyncEvent;
+import javax.servlet.AsyncListener;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import javax.ws.rs.Consumes;
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+import javax.ws.rs.QueryParam;
+import javax.ws.rs.core.Context;
+import javax.ws.rs.core.MediaType;
+import java.io.IOException;
+
+/**
+ */
+@Path("/druid-internal/v1/segments/")
+@ResourceFilters(StateResourceFilter.class)
+public class SegmentListerResource
+{
+  protected static final EmittingLogger log = new EmittingLogger(SegmentListerResource.class);
+
+  protected final ObjectMapper jsonMapper;
+  protected final ObjectMapper smileMapper;
+  protected final AuthConfig authConfig;
+  private final DataSegmentAnnouncer announcer;
+
+  @Inject
+  public SegmentListerResource(
+      @Json ObjectMapper jsonMapper,
+      @Smile ObjectMapper smileMapper,
+      AuthConfig authConfig,
+      DataSegmentAnnouncer announcer
```

Comment:
If you do what I wrote in a previous comment and separate the lookup management into a different implementation of `DataSegmentAnnouncer`, then have that concrete type injected here instead of the interface (if you mark it `@Nullable` Guice should be fine even if it's not bound because the http thing isn't being used)